src/dotnet/commands/dotnet-install/dotnet-install-tool/InstallToolCommand.cs

```diff
@@ -178,5 +202,30 @@ public override int Execute()
                 return 1;
             }
         }
+
+        private int Validate()
+        {
+            if (_binPath != null && _global)
+            {
+                throw new GracefulException("Cannot have global and bin-path as opinion at the same time.");
+            }
+
+            if (_configFilePath != null && !File.Exists(_configFilePath))
+            {
+                throw new GracefulException(
+                    string.Format(
+                        LocalizableStrings.NuGetConfigurationFileDoesNotExist,
+                        Path.GetFullPath(_configFilePath)));
+            }
+
+            // Prevent installation if any version of the package is installed
+            if (_toolPackageStore.GetInstalledPackages(_packageId).FirstOrDefault() != null)
+            {
+                _errorReporter.WriteLine(string.Format(LocalizableStrings.ToolAlreadyInstalled, _packageId).Red());
+                return 1;
+            }
+
+            return 0;
```

Comment:
But now that I get this far, it's confusing to have a Validate method that both throws and returns an error code.

Reply:
@peterhuene do you think we need to make everything return error? And also, why GracefulException is not suitable for giving error of dotnet tools? And if GracefulException is not a good exception, why we cannot make a different exception, instead, we need to manually decide the message, color of exception case by case?

Reply:
@wli3 `GracefulException` is fine for command-line syntax errors and should not drive all error handling because it inherently causes the command's help to be displayed.  Command help should only be displayed to a user if the error meant the user specified invalid command line syntax (an option conflicts with another, or the option's argument is invalid, etc.).  To put it another way: the command's help should only be displayed when it is *helpful*: if the error has nothing to do with what the user typed it should not be displayed.  For example, if a NuGet source was down, only that error should be displayed; displaying the command's help in addition to the actual error message confuses users into thinking the problem was with how they invoked `dotnet`.

`DotNetTopLevelCommandBase` only expects `GracefulException` to be thrown from the command.  Commands should not be throwing any other exceptions and are responsible for logging errors and returning a non-zero exit code in case of error, thus the complex error handling in these commands.

Reply:
I do agree that it is weird to either throw an exception or return non-zero in case of error.  I don't have a "clean" suggestion though, because I believe that an error due to a tool already being installed should not show command line help (the error isn't the way the command was invoked, just that the state of the system was not the expected one).

Reply:
Maybe a different exception can help? GracefulButNoHelpMessageException? Catch it is in command base. And put error message coloring, return code etc there.

Reply:
Maybe we just need to generalize GracefulException with a ShowHelp bool. e.g. can do `throw new GracefulException(message, showHelp: false)` and catcher can do `if (ex.ShowHelp) { ShowHelp() }`

Reply:
Works for me, although the default should probably be `true` so we don't have to scrub all existing uses.